src/main/java/org/broadinstitute/hellbender/utils/io/BlockCompressedIntervalStream.java

```diff
@@ -0,0 +1,626 @@
+                firstBlockMember = true;
+            }
+        }
+
+        public void close() {
+            // take care of any pending index entry, if necessary
+            if ( !firstBlockMember ) {
+                addIndexEntry();
+            }
+
+            try {
+                dos.flush(); // complete the data block
+
+                long indexPosition = bcos.getPosition(); // current position is the start of the index
+
+                // write the index entries
+                dos.writeInt(indexEntries.size());
+                for ( final IndexEntry indexEntry : indexEntries ) {
+                    indexEntry.write(dos);
+                }
+                dos.flush(); // and complete the block
+
+                // write a 0-length terminator block at the end that captures the index position
+                final byte[] emptyBlockWithIndexPointer =
+                        Arrays.copyOf(EMPTY_GZIP_BLOCK_WITH_INDEX_POINTER,
+                                EMPTY_GZIP_BLOCK_WITH_INDEX_POINTER.length);
+                for ( int idx = FILE_POINTER_OFFSET; idx != FILE_POINTER_OFFSET + 8; ++idx ) {
+                    emptyBlockWithIndexPointer[idx] = (byte)indexPosition;
+                    indexPosition >>>= 8;
+                }
+                os.write(emptyBlockWithIndexPointer);
+
+                bcos.close(false); // we've already handled the terminator block
+            } catch ( final IOException ioe ) {
+                throw new UserException("unable to add index and close " + path, ioe);
+            }
+        }
+
+        private void startBlock( final long filePosition, final T interval ) {
+            blockFilePosition = filePosition;
+            lastInterval = interval;
+            blockContig = interval.getContig();
+            blockStart = interval.getStart();
+            blockEnd = interval.getEnd();
+            firstBlockMember = false;
+        }
+
+        private void addIndexEntry() {
+            final CollatingInterval blockInterval =
+                    new CollatingInterval(dict, blockContig, blockStart, blockEnd);
+            indexEntries.add(new IndexEntry(blockInterval, blockFilePosition));
+        }
+    }
+
+    // a class for reading arbitrary objects from a block compressed stream with a self-contained index
+    // the only restriction is that you must supply a lambda that reads from a DataInputStream
+    //   to reconstitute the object.
+    public static final class Reader <T extends Feature> implements FeatureReader<T> {
+        final String path;
+        final FeatureCodec<T, Reader<T>> codec;
+        final long indexFilePointer;
+        final BlockCompressedInputStream bcis;
+        final DataInputStream dis;
+        final Header header;
+        final long dataFilePointer;
+        IntervalTree<Long> index;
+        boolean usedByIterator;
+
+        public Reader( final FeatureInput<T> inputDescriptor, final FeatureCodec<T, Reader<T>> codec ) {
```

Comment:
It would be nice if these could all route through 1 copy of the setup code instead of 3 slight variants but maybe that's not feasible.